File: packages/python/tryaii/cachelint/sequence.py
"""Sequence (multi-input) analysis: relations between consecutive requests.

For a list of inputs this answers "where does the sequence break, and why":
  - groups requests by (provider, normalized model) — caches are model-scoped,
    so a group boundary is itself a break (SPEC.md delta m)
  - computes the longest common prefix (LCP) between consecutive requests
  - classifies the relation: identical / append-only / truncated / diverged
  - attributes the divergence to a section and a likely cause (which detector
    fires around the break point in both versions)
  - predicts HIT / PARTIAL / MISS / AT_RISK, including TTL-gap checks when
    inputs carry an optional ISO `sent_at` (strict shared grammar, SPEC.md §1.6)
"""

from __future__ import annotations

from tryaii.cachelint import detectors, providers, tokenizers
from tryaii.cachelint._iso8601 import parse_ts_ms
from tryaii.cachelint._jsonutil import normalize_numbers
from tryaii.cachelint.analyzer import AnalyzedItem, _section_of

_DIVERGE_WINDOW_BACK = 40
_DIVERGE_WINDOW_FWD = 100


def _lcp_len(a: str, b: str) -> int:
    n = min(len(a), len(b))
    i = 0
    while i < n and a[i] == b[i]:
        i += 1
    return i


def _context(text: str, offset: int) -> str:
    lo = max(0, offset - _DIVERGE_WINDOW_BACK)
    hi = min(len(text), offset + _DIVERGE_WINDOW_FWD)
    return text[lo:hi].replace("\n", "\\n")


def _classify_cause(a: str, b: str, lcp: int) -> list[str]:
    """Which detector kinds fire around the divergence point, and did their VALUES change?

    A kind present in both windows with identical matched text is just nearby
    content, not the cause — only value differences (or one-sided presence)
    are reported.
    """
    values: dict[str, dict] = {}
    for label, text in (("a", a), ("b", b)):
        lo = max(0, lcp - _DIVERGE_WINDOW_BACK)
        hi = min(len(text), lcp + _DIVERGE_WINDOW_FWD)
        window = text[lo:hi]
        for f in detectors.scan(window):
            matched = window[f.start:f.end]
            values.setdefault(f.kind, {"a": set(), "b": set()})[label].add(matched)
    causes = []
    for kind, sides in sorted(values.items()):
        if sides["a"] and sides["b"]:
            if sides["a"] != sides["b"]:
                causes.append(f"{kind} value changed")
            # identical values on both sides: nearby content, not the cause
        elif sides["b"]:
            causes.append(f"{kind} introduced")
        elif sides["a"]:
            causes.append(f"{kind} removed")
    return causes


def _pair(prev: AnalyzedItem, curr: AnalyzedItem) -> dict:
    a, b = prev.canonical, curr.canonical
    provider_key = prev.resolved["provider_key"]
    upstream = prev.resolved.get("upstream")
    min_tokens = prev.resolved["min_tokens"]
    spec = prev.resolved["spec"]

    lcp = _lcp_len(a, b)
    lcp_tokens = tokenizers.count_tokens(a[:lcp], provider_key, upstream).tokens
    notes: list[str] = []
    diverged_at = None
    causes: list[str] = []

    if a == b:
        relation = "identical"
    elif lcp == len(a):
        relation = "append-only"       # b extends a: the classic growing conversation
    elif lcp == len(b):
        relation = "truncated"         # b is a prefix of a: history was trimmed
        notes.append("The newer request is a PREFIX of the older one — history trimming? "
                     "Its full length can still hit the cache, but trimming mid-conversation "
                     "usually signals a context-window strategy that fights caching.")
    else:
        relation = "diverged"
        sec = _section_of(prev.sections, min(lcp, max(0, len(a) - 1)))
        causes = _classify_cause(a, b, lcp)
        diverged_at = {
            "offset": lcp,
            "section": sec.name if sec else "?",
            "section_offset": (lcp - sec.start) if sec else None,
            "context_prev": _context(a, lcp),
            "context_curr": _context(b, lcp),
        }
        if sec and sec.name == "tools":
            notes.append("Divergence inside the TOOLS section — tool changes invalidate the "
                         "ENTIRE cache (tools render first).")
        elif sec and sec.name == "system":
            notes.append("Divergence inside the SYSTEM section — invalidates system + messages "
                         "(tools may survive on Anthropic).")

    # expected outcome
    threshold_ok = (min_tokens is None) or (lcp_tokens >= min_tokens)
    if relation in ("identical", "append-only"):
        expected = "HIT" if threshold_ok else "MISS"
        expected_cached = lcp_tokens if threshold_ok else 0
        if not threshold_ok:
            notes.append(f"Common prefix ~{lcp_tokens} tok is below the {min_tokens} minimum — "
                         "no cache credit despite the clean prefix.")
    elif relation == "truncated":
        expected = "HIT" if threshold_ok else "MISS"
        expected_cached = lcp_tokens if threshold_ok else 0
    else:
        if threshold_ok and min_tokens is not None:
            expected = "PARTIAL"
            expected_cached = lcp_tokens
            notes.append(f"Only the ~{lcp_tokens}-tok common prefix gets cache credit; "
                         "everything after the divergence re-processes at full price.")
        elif min_tokens is None:
            expected = "UNKNOWN"
            expected_cached = None
        else:
            expected = "MISS"
            expected_cached = 0
            notes.append(f"Common prefix ~{lcp_tokens} tok < {min_tokens} minimum — "
                         "the divergence effectively kills caching for this pair.")

    # TTL gap check (SPEC.md §1.6: strict shared grammar, epoch-ms arithmetic)
    ttl_check = None
    t_prev, t_curr = parse_ts_ms(prev.sent_at), parse_ts_ms(curr.sent_at)
    if t_prev is not None and t_curr is not None:
        gap = (t_curr - t_prev) / 1000.0
        ttl_check = {"gap_seconds": normalize_numbers(round(gap, 1)),
                     "ttl_seconds": spec.ttl_seconds}
        if spec.ttl_seconds is None:
            ttl_check["note"] = ("No TTL guarantee / unpublished lifetime for this provider — "
                                 "any gap is at-risk.")
            if expected == "HIT":
                expected = "AT_RISK"
        elif gap > spec.ttl_seconds and expected in ("HIT", "PARTIAL"):
            if provider_key == "openai" and gap <= 24 * 3600:
                expected = "AT_RISK"
                ttl_check["note"] = (f"Gap {gap:.0f}s exceeds the 5-10 min sliding window; may "
                                     "still hit via 24h extended retention (not guaranteed).")
            else:
                expected = "MISS"
                ttl_check["note"] = (f"Gap {gap:.0f}s exceeds the default TTL "
                                     f"({spec.ttl_seconds}s) — entry likely evicted. "
                                     "Consider the 1h TTL / keep-alive pings where available.")

    if spec.enablement == "explicit" and expected in ("HIT", "PARTIAL", "AT_RISK"):
        notes.append("Explicit-enablement provider: the hit requires a cache breakpoint "
                     "placed at/inside the common prefix on BOTH requests.")

    return {
        "relation": relation,
        "lcp_chars": lcp,
        "lcp_tokens": lcp_tokens,
        "expected": expected,
        "expected_cached_tokens": expected_cached,
        "diverged_at": diverged_at,
        "likely_causes": causes,
        "ttl_check": ttl_check,
        "notes": notes,
    }


def analyze_sequences(items: list[AnalyzedItem]) -> list[dict]:
    """Group by (provider, normalized model) and analyze consecutive-pair relations."""
    groups: dict[tuple, list] = {}
    order: list[tuple] = []
    for it in items:
        # SPEC.md delta m: normalize the model for grouping; the group header
        # still displays the first-seen raw string.
        key = (it.resolved["provider_key"], providers.norm_model(it.data["model"]))
        if key not in groups:
            groups[key] = []
            order.append(key)
        groups[key].append(it)

    results = []
    for key in order:
        members = groups[key]
        provider_key = key[0]
        model = members[0].data["model"]
        spec = members[0].resolved["spec"]
        min_tokens = members[0].resolved["min_tokens"]

        group_result: dict = {
            "group": {"provider": spec.display, "provider_key": provider_key, "model": model},
            "count": len(members),
            "indices": [m.data["index"] for m in members],
            "pairs": [],
            "shared_prefix_all": None,
            "summary": "",
        }

        if len(members) == 1:
            group_result["summary"] = (
                "Single request — caching needs >= 2 same-prefix requests inside the TTL window "
                "to pay off (a lone write costs the premium for nothing where one exists).")
            results.append(group_result)
            continue

        # LCP across ALL members = the fleet-wide shared cache unit
        shared = members[0].canonical
        shared_len = len(shared)
        for m in members[1:]:
            shared_len = min(shared_len, _lcp_len(shared[:shared_len], m.canonical))
        shared_tokens = tokenizers.count_tokens(
            members[0].canonical[:shared_len],
            provider_key, members[0].resolved.get("upstream")).tokens
        group_result["shared_prefix_all"] = {
            "tokens": shared_tokens,
            "meets_threshold": (min_tokens is None) or (shared_tokens >= min_tokens),
            "min_tokens": min_tokens,
        }

        hits = 0
        for prev, curr in zip(members, members[1:]):
            pair = _pair(prev, curr)
            pair["from"] = prev.data["index"]
            pair["to"] = curr.data["index"]
            group_result["pairs"].append(pair)
            if pair["expected"] in ("HIT", "PARTIAL"):
                hits += 1

        n_pairs = len(group_result["pairs"])
        ok = "clears" if group_result["shared_prefix_all"]["meets_threshold"] else "does NOT clear"
        group_result["summary"] = (
            f"{len(members)} requests; {hits}/{n_pairs} consecutive transitions expected to get "
            f"cache credit. Shared prefix across all requests: ~{shared_tokens} tok, which {ok} "
            f"the activation threshold"
            + (f" ({min_tokens})." if min_tokens is not None else " (unpublished threshold).")
        )
        results.append(group_result)

    if len(order) > 1:
        results.append({
            "group": {"provider": "-", "provider_key": "-", "model": "-"},
            "count": 0, "indices": [], "pairs": [], "shared_prefix_all": None,
            "summary": ("NOTE: the input list spans multiple (provider, model) pairs — caches "
                        "are model- and provider-scoped, so requests in different groups never "
                        "share cache. A model/provider switch mid-sequence = cold cache."),
        })
    return results

File: packages/node/src/cachelint/sequence.ts
/**
 * Sequence (multi-input) analysis: relations between consecutive requests.
 * TypeScript mirror of the Python reference — grouping by (provider,
 * normalized model) (SPEC.md delta m), code-point LCPs (SPEC.md §1.1), TTL
 * gaps via the strict shared ISO-8601 grammar (SPEC.md §1.6).
 */

import { scan } from './detectors.js';
import { normModel } from './providers.js';
import { countTokens } from './tokenizers.js';
import { CpIndex, cpLcp } from './util/codepoints.js';
import { formatFixed, halfEvenRound } from './util/halfEven.js';
import { parseTsMs } from './util/iso8601.js';
import { sectionOf, type AnalyzedItem } from './analyzer.js';

const DIVERGE_WINDOW_BACK = 40;
const DIVERGE_WINDOW_FWD = 100;

function context(cp: CpIndex, offset: number): string {
  const lo = Math.max(0, offset - DIVERGE_WINDOW_BACK);
  const hi = Math.min(cp.length, offset + DIVERGE_WINDOW_FWD);
  return cp.slice(lo, hi).split('\n').join('\\n');
}

function setsEqual(a: Set<string>, b: Set<string>): boolean {
  if (a.size !== b.size) return false;
  for (const v of a) if (!b.has(v)) return false;
  return true;
}

/**
 * Which detector kinds fire around the divergence point, and did their VALUES
 * change? A kind present in both windows with identical matched text is just
 * nearby content, not the cause.
 */
function classifyCause(a: string, b: string, lcp: number): string[] {
  const values = new Map<string, { a: Set<string>; b: Set<string> }>();
  for (const [label, text] of [
    ['a', a],
    ['b', b],
  ] as const) {
    const cp = new CpIndex(text);
    const lo = Math.max(0, lcp - DIVERGE_WINDOW_BACK);
    const hi = Math.min(cp.length, lcp + DIVERGE_WINDOW_FWD);
    const window = cp.slice(lo, hi);
    const wcp = new CpIndex(window);
    for (const f of scan(window)) {
      const matched = wcp.slice(f.start, f.end);
      if (!values.has(f.kind)) values.set(f.kind, { a: new Set(), b: new Set() });
      (values.get(f.kind) as { a: Set<string>; b: Set<string> })[label].add(matched);
    }
  }
  const causes: string[] = [];
  for (const kind of [...values.keys()].sort()) {
    const sides = values.get(kind) as { a: Set<string>; b: Set<string> };
    if (sides.a.size && sides.b.size) {
      if (!setsEqual(sides.a, sides.b)) causes.push(`${kind} value changed`);
      // identical values on both sides: nearby content, not the cause
    } else if (sides.b.size) {
      causes.push(`${kind} introduced`);
    } else if (sides.a.size) {
      causes.push(`${kind} removed`);
    }
  }
  return causes;
}

function pairAnalysis(prev: AnalyzedItem, curr: AnalyzedItem): Record<string, unknown> {
  const a = prev.canonical;
  const b = curr.canonical;
  const providerKey = prev.resolved.provider_key;
  const upstream = prev.resolved.upstream;
  const minTokens = prev.resolved.min_tokens;
  const spec = prev.resolved.spec;
  const aCp = new CpIndex(a);
  const bCp = new CpIndex(b);

  const lcp = cpLcp(a, b);
  const lcpTokens = countTokens(aCp.slice(0, lcp), providerKey, upstream).tokens;
  const notes: string[] = [];
  let divergedAt: Record<string, unknown> | null = null;
  let causes: string[] = [];

  let relation: string;
  if (a === b) {
    relation = 'identical';
  } else if (lcp === aCp.length) {
    relation = 'append-only'; // b extends a: the classic growing conversation
  } else if (lcp === bCp.length) {
    relation = 'truncated'; // b is a prefix of a: history was trimmed
    notes.push(
      'The newer request is a PREFIX of the older one — history trimming? ' +
        'Its full length can still hit the cache, but trimming mid-conversation ' +
        'usually signals a context-window strategy that fights caching.',
    );
  } else {
    relation = 'diverged';
    const sec = sectionOf(prev.sections, Math.min(lcp, Math.max(0, aCp.length - 1)));
    causes = classifyCause(a, b, lcp);
    divergedAt = {
      offset: lcp,
      section: sec ? sec.name : '?',
      section_offset: sec ? lcp - sec.start : null,
      context_prev: context(aCp, lcp),
      context_curr: context(bCp, lcp),
    };
    if (sec && sec.name === 'tools') {
      notes.push(
        'Divergence inside the TOOLS section — tool changes invalidate the ' +
          'ENTIRE cache (tools render first).',
      );
    } else if (sec && sec.name === 'system') {
      notes.push(
        'Divergence inside the SYSTEM section — invalidates system + messages ' +
          '(tools may survive on Anthropic).',
      );
    }
  }

  // expected outcome
  const thresholdOk = minTokens === null || lcpTokens >= minTokens;
  let expected: string;
  let expectedCached: number | null;
  if (relation === 'identical' || relation === 'append-only') {
    expected = thresholdOk ? 'HIT' : 'MISS';
    expectedCached = thresholdOk ? lcpTokens : 0;
    if (!thresholdOk) {
      notes.push(
        `Common prefix ~${lcpTokens} tok is below the ${minTokens} minimum — ` +
          'no cache credit despite the clean prefix.',
      );
    }
  } else if (relation === 'truncated') {
    expected = thresholdOk ? 'HIT' : 'MISS';
    expectedCached = thresholdOk ? lcpTokens : 0;
  } else {
    if (thresholdOk && minTokens !== null) {
      expected = 'PARTIAL';
      expectedCached = lcpTokens;
      notes.push(
        `Only the ~${lcpTokens}-tok common prefix gets cache credit; ` +
          'everything after the divergence re-processes at full price.',
      );
    } else if (minTokens === null) {
      expected = 'UNKNOWN';
      expectedCached = null;
    } else {
      expected = 'MISS';
      expectedCached = 0;
      notes.push(
        `Common prefix ~${lcpTokens} tok < ${minTokens} minimum — ` +
          'the divergence effectively kills caching for this pair.',
      );
    }
  }

  // TTL gap check (SPEC.md §1.6: strict shared grammar, epoch-ms arithmetic)
  let ttlCheck: Record<string, unknown> | null = null;
  const tPrev = parseTsMs(prev.sent_at);
  const tCurr = parseTsMs(curr.sent_at);
  if (tPrev !== null && tCurr !== null) {
    const gap = (tCurr - tPrev) / 1000.0;
    ttlCheck = { gap_seconds: halfEvenRound(gap, 1), ttl_seconds: spec.ttl_seconds };
    if (spec.ttl_seconds === null) {
      ttlCheck.note =
        'No TTL guarantee / unpublished lifetime for this provider — any gap is at-risk.';
      if (expected === 'HIT') expected = 'AT_RISK';
    } else if (gap > spec.ttl_seconds && (expected === 'HIT' || expected === 'PARTIAL')) {
      if (providerKey === 'openai' && gap <= 24 * 3600) {
        expected = 'AT_RISK';
        ttlCheck.note =
          `Gap ${formatFixed(gap, 0)}s exceeds the 5-10 min sliding window; may ` +
          'still hit via 24h extended retention (not guaranteed).';
      } else {
        expected = 'MISS';
        ttlCheck.note =
          `Gap ${formatFixed(gap, 0)}s exceeds the default TTL ` +
          `(${spec.ttl_seconds}s) — entry likely evicted. ` +
          'Consider the 1h TTL / keep-alive pings where available.';
      }
    }
  }

  if (spec.enablement === 'explicit' && ['HIT', 'PARTIAL', 'AT_RISK'].includes(expected)) {
    notes.push(
      'Explicit-enablement provider: the hit requires a cache breakpoint ' +
        'placed at/inside the common prefix on BOTH requests.',
    );
  }

  return {
    relation,
    lcp_chars: lcp,
    lcp_tokens: lcpTokens,
    expected,
    expected_cached_tokens: expectedCached,
    diverged_at: divergedAt,
    likely_causes: causes,
    ttl_check: ttlCheck,
    notes,
  };
}

/** Group by (provider, normalized model) and analyze consecutive-pair relations. */
export function analyzeSequences(items: AnalyzedItem[]): Record<string, unknown>[] {
  const groups = new Map<string, AnalyzedItem[]>();
  for (const it of items) {
    // SPEC.md delta m: normalize the model for grouping; the group header
    // still displays the first-seen raw string.
    const key = JSON.stringify([it.resolved.provider_key, normModel(it.data.model as string)]);
    if (!groups.has(key)) groups.set(key, []);
    (groups.get(key) as AnalyzedItem[]).push(it);
  }

  const results: Record<string, unknown>[] = [];
  for (const members of groups.values()) {
    const providerKey = members[0].resolved.provider_key;
    const model = members[0].data.model as string;
    const spec = members[0].resolved.spec;
    const minTokens = members[0].resolved.min_tokens;

    const groupResult: Record<string, unknown> = {
      group: { provider: spec.display, provider_key: providerKey, model },
      count: members.length,
      indices: members.map((m) => m.data.index),
      pairs: [] as Record<string, unknown>[],
      shared_prefix_all: null as Record<string, unknown> | null,
      summary: '',
    };

    if (members.length === 1) {
      groupResult.summary =
        'Single request — caching needs >= 2 same-prefix requests inside the TTL window ' +
        'to pay off (a lone write costs the premium for nothing where one exists).';
      results.push(groupResult);
      continue;
    }

    // LCP across ALL members = the fleet-wide shared cache unit
    const firstCp = new CpIndex(members[0].canonical);
    let sharedLen = firstCp.length;
    for (const m of members.slice(1)) {
      sharedLen = Math.min(sharedLen, cpLcp(firstCp.slice(0, sharedLen), m.canonical));
    }
    const sharedTokens = countTokens(
      firstCp.slice(0, sharedLen),
      providerKey,
      members[0].resolved.upstream,
    ).tokens;
    groupResult.shared_prefix_all = {
      tokens: sharedTokens,
      meets_threshold: minTokens === null || sharedTokens >= minTokens,
      min_tokens: minTokens,
    };

    let hits = 0;
    const pairs = groupResult.pairs as Record<string, unknown>[];
    for (let i = 1; i < members.length; i++) {
      const pair = pairAnalysis(members[i - 1], members[i]);
      pair.from = members[i - 1].data.index;
      pair.to = members[i].data.index;
      pairs.push(pair);
      if (pair.expected === 'HIT' || pair.expected === 'PARTIAL') hits++;
    }

    const nPairs = pairs.length;
    const ok = (groupResult.shared_prefix_all as Record<string, unknown>).meets_threshold
      ? 'clears'
      : 'does NOT clear';
    groupResult.summary =
      `${members.length} requests; ${hits}/${nPairs} consecutive transitions expected to get ` +
      `cache credit. Shared prefix across all requests: ~${sharedTokens} tok, which ${ok} ` +
      'the activation threshold' +
      (minTokens !== null ? ` (${minTokens}).` : ' (unpublished threshold).');
    results.push(groupResult);
  }

  if (groups.size > 1) {
    results.push({
      group: { provider: '-', provider_key: '-', model: '-' },
      count: 0,
      indices: [],
      pairs: [],
      shared_prefix_all: null,
      summary:
        'NOTE: the input list spans multiple (provider, model) pairs — caches ' +
        'are model- and provider-scoped, so requests in different groups never ' +
        'share cache. A model/provider switch mid-sequence = cold cache.',
    });
  }
  return results;
}
